fauna/wire_protocol.py

```diff
@@ -260,6 +289,34 @@ def _decode_dict(dct: dict, object_tagged: bool):
             if "@date" in dct:
                 return parse_date(dct["@date"]).date()
             if "@doc" in dct:
-                return DocumentReference.from_string(dct["@doc"])
+                value = dct["@doc"]
+                if isinstance(value, str):
+                    # Not distinguishing between DocumentReference and NamedDocumentReference because this shouldn't
+                    # be an issue much longer
+                    return DocumentReference.from_string(value)
+
+                contents = FaunaDecoder._decode(value)
+
+                if "id" in value:
+                    return Document(contents)
+                elif "name" in value:
+                    return NamedDocument(contents)
+                else:
+                    # Unsupported document reference. Return the unwrapped value to futureproof.
+                    return contents
+
+            if "@ref" in dct:
+                value = dct["@ref"]
+                col = FaunaDecoder._decode(value["coll"])
+
+                if "id" in value:
+                    return DocumentReference(col, value["id"])
```

Comment:
Also needs str -> int conversion. Doing this in the `DocumentReference` constructor would probably be best.

Reply:
I'm going to go back to storing this as a string